### ISSUE TITLE
clightgen generates Coq code that doesn't compile because of unsanitized names

### ISSUE THREAD

**Issue by joom** [AbsInt/CompCert#319]
Take the following C program:

```c
int a$b() {
  return 0;
}

int main(int argc, const char *argv[])
{
  return a$b();
}
```

Now, if we compile this with clightgen, the generated Coq file contains these:

```coq
Definition _a_b : ident := 52%positive.

Definition f_a$b := {|
  fn_return := tint;
  fn_callconv := cc_default;
  fn_params := nil;
  fn_vars := nil;
  fn_temps := nil;
  fn_body := (Sreturn (Some (Econst_int (Int.repr 0) tint)))
|}.
```

The problem is that `$` is not a valid symbol in Coq identifiers, so trying to compile the Coq file gives us a lexer error because we have a definition named `f_a$b`.

Also interestingly, the `ident` created for this function *is* sanitized and it doesn't contain `$`. But then this creates another problem, take the following C program:

```c
int a$b() {
  return 0;
}

int a_b() {
  return 0;
}

int main(int argc, const char *argv[])
{
  return a$b() + a_b();
}
```

Now clightgen creates two `ident` definitions, both named `_a_b`. This time the generated Coq file doesn't compile because there are two definitions with the same name.

If these programs are supposed to fail, there should at least be a warning about these in clightgen.

I get that preserving original C function names matters there's a naming convention (idents start with `_`, function definitions start with `f_` and such) in the generated Coq files and tools that use Clight take advantage of this convention. But then why are the `ident` names sanitized and `function` definition names are not?

If we choose to sanitize `function` definition names as well, that should be an easy change in the `define_idents` function in `exportclight/ExportClight.ml`. But whether both sanitized or unsanitized, a warning for this would be nice.

**Comment by xavierleroy** [comment]
Yes, function names and global variable names were not "sanitized" consistently.  Commit 8fbe9863 fixes this.

As to `a$b` and `a_b` "sanitizing" to the same Coq identifier: this is a known limitation of clightgen.  Nothing serious here: it should happen very rarely and results in a Coq error later (as opposed to silently producing a wrong AST).  We could fix it by using a well-chosen Unicode letter to encode `$`.

**Comment by xavierleroy** [comment]
Closing this issue, as it seems to be fixed by 8fbe986.
